pkg/apis/core/validation/validation.go

```diff
@@ -3720,10 +3756,69 @@ func ValidatePodCreate(pod *core.Pod, opts PodValidationOptions) field.ErrorList
 	if len(pod.Spec.EphemeralContainers) > 0 {
 		allErrs = append(allErrs, field.Forbidden(fldPath.Child("ephemeralContainers"), "cannot be set on create"))
 	}
+	allErrs = append(allErrs, ValidateSeccompAnnotationsAndFields(pod.ObjectMeta, &pod.Spec, field.NewPath(""))...)
+
+	return allErrs
+}
+
+// ValidateSeccompAnnotationsAndFields iterates through all containers and ensure that when both seccompProfile and seccomp annotations exist they match.
+func ValidateSeccompAnnotationsAndFields(objectMeta metav1.ObjectMeta, podSpec *core.PodSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if podSpec.SecurityContext != nil && podSpec.SecurityContext.SeccompProfile != nil {
+		// If both seccomp annotations and fields are specified, the values must match.
+		if annotation, found := objectMeta.Annotations[v1.SeccompPodAnnotationKey]; found {
+			fldPath = fldPath.Child("spec").Child("securityContext").Child("seccompProfile")
+			allErrs = append(allErrs, validateSeccompAnnotationsAndFieldsMatch(annotation, podSpec.SecurityContext.SeccompProfile, fldPath))
+		}
+	}
+
+	podshelper.VisitContainersWithPath(podSpec, func(c *core.Container, fldPath *field.Path) bool {
```

Comment:
interestingly, the fldPath passed into ValidateSeccompAnnotationsAndFields is lost here... that becomes important when we're validating a pod spec template inside an object like a deployment... the field path to an error there should be something like `spec.template.spec.securityContext.seccompProfile`, etc

Reply:
looks like we need a test that exercises initContainers as well to make sure the field path is correct

Reply:
added this unit test to assert all container types:

```
			"all container types with template pod path",
			field.NewPath("template", "spec"),
			&api.PodSpec{
				Containers: []api.Container{
					{Name: "c1"},
					{Name: "c2"},
				},
				InitContainers: []api.Container{
					{Name: "i1"},
					{Name: "i2"},
				},
				EphemeralContainers: []api.EphemeralContainer{
					{EphemeralContainerCommon: api.EphemeralContainerCommon{Name: "e1"}},
					{EphemeralContainerCommon: api.EphemeralContainerCommon{Name: "e2"}},
				},
			},
			[]string{"template.spec.initContainers[0]", "template.spec.initContainers[1]", "template.spec.containers[0]", "template.spec.containers[1]", "template.spec.ephemeralContainers[0]", "template.spec.ephemeralContainers[1]"},
```